src/msg/async/ProtocolV1.cc

```diff
@@ -1304,7 +1304,8 @@ Message *ProtocolV1::_get_next_outgoing(bufferlist *bl) {
     ceph_assert(!it->second.empty());
     list<pair<bufferlist, Message *> >::iterator p = it->second.begin();
     m = p->second;
-    if (bl) bl->swap(p->first);
+    if (p->first.length() && bl)
+      bl->swap(p->first);
```

Comment:
hmm, this will assume *bl is empty

Reply:
the only caller side is

https://github.com/ceph/ceph/blob/2499641c7212da02a020dbbac2933f0773991cfe/src/msg/async/ProtocolV1.cc#L317-L318

so i guess it's fine. or we could add an `assert()` to make this assumption more visible.

Reply:
@tchaikov . i'll add assert in this.   BTW  Does it need to add assert(!(_len == 0 && other_len == 0)) in bufferlist::swap?

Reply:
> Does it need to add assert(!(_len == 0 && other_len == 0)) in bufferlist::swap?

why would we need to do so?

Reply:
check other place met the same problem which two empty bufferlist do swap.

Reply:
i'd suggest s/ceph_assert/assert/ to avoid the overhead in production env.

Reply:
> check other place met the same problem which two empty bufferlist do swap.

why do you think it's a bug?

Reply:
avoid meaningless swap. Yes, it's not a bug.

Reply:
i don't think we should `assert()` in this case then, unless we audit all caller sites, and claim that swapping two empty bufferlist should not be tolerated.

Reply:
got it. @tchaikov

Reply:
@majianpeng  in case you missed my comment. i suggest  s/ceph_assert/assert/

Reply:
@tchaikov . What's the different between ceph_assert and assert?

Reply:
ceph_assert() won't be optimized out if NDEBUG is defined, while assert() will. please refer `man assert` and `src/include/ceph_assert.h`.

Reply:
thanks to your detailed explanation @tchaikov